.shiny_ui <- function()
{
  shinydashboard::dashboardPage(
    shinydashboard::dashboardHeader(title = "inspectorORF"),
    shinydashboard::dashboardSidebar(
      shinydashboard::sidebarMenu(
        id = "tabs",
        shinydashboard::menuItem("Input", tabName = "input", icon = shiny::icon("upload")),
        shinydashboard::menuItem("Data", tabName = "data", icon = shiny::icon("table")),
        shinydashboard::menuItem("Vignette", tabName = "vignette", icon = shiny::icon("book")),
        shinydashboard::menuItem("About", tabName = "about", icon = shiny::icon("address-card"))
      )
    ),
    shinydashboard::dashboardBody(
      shinydashboard::tabItems(
        shinydashboard::tabItem(tabName = "input",
          shiny::fluidRow(
            shinydashboard::box(
              title = "Input data",
              status = "primary",
              width = 12,

              shiny::h4(shiny::strong("Step 1"), " - Select input type"),
              shiny::p("'Quantified data' requires RNA-Seq bed files quantified with bedtools and P-Site data created by calculating as such with ORFquant."),
              shiny::p("'Processed inspectorORF bed' requires previously processed bed tracks at the gene or transcript level previously generated by inspectorORF."),

              shiny::radioButtons(
                inputId = "input_type",
                label = "Select input type:",
                choices = c("Quantified data" = "raw", "Processed inspectorORF bed" = "bed"),
                selected = "raw",
                inline = TRUE
              ),

              shiny::conditionalPanel(
                condition = "input.input_type == 'bed'",
                shiny::actionButton("bed_example_data", "Example data"),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 2"), " - Select BED file generated by inspectorORF"),
                shiny::fileInput(
                  "bed_file",
                  "BED File",
                  multiple = FALSE,
                  accept = c(".bed", ".bed.gz")
                ),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 3"), " - Select Genomic annotation and 2bit files"),
                shiny::fileInput(
                  "gtf_file",
                  "Genomic GTF file",
                  multiple = FALSE,
                  accept = c(".gtf")
                ),
                shiny::fileInput(
                  "sequence_file",
                  "Genomic 2BIT file",
                  multiple = FALSE,
                  accept = c(".2bit")
                )
              ),

              shiny::conditionalPanel(
                condition = "input.input_type == 'raw'",
                shiny::actionButton("raw_example_data", "Example data"),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 2"), " - Select RNA-Seq data"),
                shiny::fileInput(
                  "rnaseq_file",
                  "Quantified RNA-Seq data",
                  multiple = FALSE,
                  accept = c(".bed", ".bed.gz")
                ),
                shiny::uiOutput("rnaseq_name"),
                shiny::actionButton("add_rnaseq", "Add RNA-Seq file"),
                shiny::br(),
                shiny::br(),
                DT::DTOutput("rnaseq_table"),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 3"), " - Select P-Site data"),
                shiny::fileInput(
                  "riboseq_file",
                  "P-Sites calculated by ORFquant",
                  multiple = FALSE,
                ),
                shiny::uiOutput("riboseq_name"),
                shiny::actionButton("add_orfquant", "Add P-Sites file"),
                shiny::br(),
                shiny::br(),
                DT::DTOutput("orfquant_table"),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 4"), " - Select Genomic annotation and 2bit files"),
                shiny::fileInput(
                  "gtf_file",
                  "Genomic GTF file",
                  multiple = FALSE,
                  accept = c(".gtf")
                ),
                shiny::fileInput(
                  "sequence_file",
                  "Genomic 2BIT file",
                  multiple = FALSE,
                  accept = c(".2bit")
                ),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 5"), " - Processe data for gene or transcript level"),
                shiny::textAreaInput(
                  inputId = "id_input",
                  label = "Enter gene IDs or transcript IDs (separated by new line):",
                  value = "",
                  placeholder = "",
                  rows = 10,
                  cols = 40
                ),
                shiny::actionButton("process_tracks", "Generate tracks"),
                shiny::hr(),
                shiny::h4(shiny::strong("Step 6 (Optional)"), " - Export processed data for faster, future importing"),
                shiny::downloadButton(
                  outputId = "export_tracks",
                  label = "Export tracks"
                ),
              ),
            ),
            # shinydashboard::box(
            #   title = "Instructions",
            #   shiny::p("For full detailed instructions, please see the 'Vignette' tab."),
            #   shiny::h4(shiny::strong("Step 1"), " - Selecting the input type"),
            #   shiny::p("For first time use, RNA-Seq and ORFquant output data is required to be processed with the 'Quantified data' option."),
            #   shiny::p("Upon saving the processed information as a bed file, future usage can be performed with 'processed inspectorORF bed'."),
            #   shiny::conditionalPanel(
            #     condition = "input.input_type == 'bed'",
            #     shiny::h4(shiny::strong("Step 2"), " - Importing BED file"),
            #     shiny::p("inspectorORF expects the data within the BED file to be alternating P-site reads and RNA-Seq reads for each chromosome position."),
            #     shiny::p("This is automatically generated by inspectorORF but please see <LINK HERE> for full details if you wish to create the file manually.."),
            #     shiny::h4(shiny::strong("Step 3"), " - Selecting BED file track type"),
            #     shiny::p("Bed files consisting of gene level or track level information can be imported."),
            #     shiny::p("There are advantages to both. For gene level, any transcript isoform of a gene can be recreated during visualisation, whereas transcript level files can be significantly smaller if you're working with a lot of candidates."),
            #     shiny::h4(shiny::strong("Step 4"), " - Importing genomic annotation files"),
            #     shiny::p("2bit files can be generated from fasta files with the UCSC tool 'faToTwoBit'."),

            #     shiny::tags$pre(
            #       shiny::tags$code(
            #         style = "
            #           display: block;
            #           font-family: monospace;
            #           white-space: pre-wrap;
            #           text-indent: 0;
            #         ",
            #         shiny::HTML("$ faToTwoBit genome.fasta genome.2bit")
            #       )
            #     ),
            #     shiny::p("Please see",
            #       shiny::a("https://genome.ucsc.edu/goldenPath/help/twoBit.html", href = "https://genome.ucsc.edu/goldenPath/help/twoBit.html", target="_blank"),
            #       "for further details.")
            #   ),

            #   shiny::conditionalPanel(
            #     condition = "input.input_type == 'raw'",
            #     shiny::h4(shiny::strong("Step 2"), " - Importing RNA-Seq data"),
            #     shiny::p("Quantified RNA-Seq coverage can be calculated with bedtools with the following commands:"),
            #     shiny::tags$pre(
            #       shiny::tags$code(
            #         style = "
            #           display: block;
            #           font-family: monospace;
            #           white-space: pre-wrap;
            #           text-indent: 0;
            #         ",
            #         shiny::HTML(paste("$ bedtools coverage -abam alignment.bam -b annotation.gtf > rna_tracks.bed",
            #           "$ gzip rna_tracks.bed # optional, but recommended for large files",
            #           sep = "<br>"
            #         ))
            #       )
            #     ),
            #     shiny::p(shiny::strong("NOTE"), "- At least one RNA-Seq coverage .bed(.gz) file must be imported."),
            #     shiny::h4(shiny::strong("Step 3"), " - Importing Ribo-Seq data"),
            #     shiny::p("Requires the for_ORFquant file generated by ORFquant with the '' function."),
            #     shiny::p(shiny::strong("NOTE"), "- At least one P-Site file must be imported."),
            #     shiny::p(shiny::strong("Additional notes"), " - Gene level information for additional replicates, conditions or further data from for example RNA-Seq, Ribo-Seq or other sequencing methods such as long-read sequencing can be imported."),
            #     shiny::h4(shiny::strong("Step 4"), " - Importing genomic annotation files"),
            #     shiny::p("2bit files can be generated from fasta files with the UCSC tool 'faToTwoBit'."),

            #     shiny::tags$pre(
            #       shiny::tags$code(
            #         style = "
            #           display: block;
            #           font-family: monospace;
            #           white-space: pre-wrap;
            #           text-indent: 0;
            #         ",
            #         shiny::HTML("$ faToTwoBit genome.fasta genome.2bit")
            #       )
            #     ),
            #     shiny::p("Please see",
            #       shiny::a("https://genome.ucsc.edu/goldenPath/help/twoBit.html", href = "https://genome.ucsc.edu/goldenPath/help/twoBit.html", target="_blank"),
            #       "for further details."),
            #     shiny::h4(shiny::strong("Step 5"), " - Generate tracks"),
            #     shiny::p("Generate tracks at the gene or transcript level."),
            #     shiny::p(shiny::strong("NOTE"), " - This can take some type if the GTF file contains a large genome, especially when generating information at the gene level."),
            #     shiny::h4(shiny::strong("Optional Step 6"), " - Importing genomic annotation files"),
            #     shiny::p("Finally, processed data can be exported as gene level or transcript level track information for easier and faster importing in the future.")
            #   ),
            #   shiny::br(),
            #   shiny::br(),
            #   shiny::p("Once files have been imported and processed, plots can be manipulated within the 'Data' tab.")
            # ),
            # shinydashboard::box(
            #   title = "Input metrics",
            #   # shiny::h4("BED details"),
            #   # reactable::reactableOutput("bed_stats"),
            #   shiny::h4("GTF details"),
            #   reactable::reactableOutput("gtf_stats")
            # )
          )
        ),
        shinydashboard::tabItem(
          tabName = "data",
          shiny::fluidRow(
            shiny::includeCSS(system.file("www/stylesheet.css", package = "inspectorORF")),
            shinydashboard::box(
              title = "Bed data",
              width = 12,
              collapsible = TRUE,
              reactable::reactableOutput("transcript_id_table")
            ),
            shinydashboard::box(
              title = "ORFs",
              width = 12,
              reactable::reactableOutput("orf_id_table")
            ),
            shinydashboard::box(
              title = "Plot data",
              width = 12,
              collapsible = TRUE,
              shiny::fluidRow(
                shiny::column(
                  width = 12,
                  shiny::selectInput("plot_data", "Plot Data", choices = c(""), multiple = TRUE),
                )
              )
            ),
            shinydashboard::box(
              title = "Plot region",
              width = 6,
              collapsible = TRUE,
              shiny::fluidRow(
                shiny::column(
                  width = 12,
                  shiny::selectInput(
                    "plot_region",
                    "Plot region",
                    choices = c(
                      "Full transcript",
                      "ORF Only",
                      "Upstream 5' + ORF",
                      "ORF + downstream 3'",
                      "Upstream 5' + ORF + downstream 3'"
                    )
                  ),
                  shiny::checkboxInput('scale_plot', 'Scale to max P-sites'),
                  shiny::checkboxInput('transcript_summary', 'Summarise rest of transcript P-sites'),
                ),
              ),
            ),
            shinydashboard::box(
              title = "Annotations",
              width = 6,
              collapsible = TRUE,
              shiny::fluidRow(
                shiny::column(
                  width = 12,
                  shiny::checkboxInput('annotate_start', 'Annotate start codon'),
                  shiny::selectInput("annotate_codons", "Annotate additional codons", choices = c("AUG", "CUG", "GUG", "UUG", "AAG", "ACG"), multiple = T),
                  shiny::checkboxInput('annotate_in_frame', 'Annotate in-frame only')
                ),
              ),
            ),
            shinydashboard::box(
              title = "Plot",
              width = 12,
              shiny::fluidRow(
                shiny::column(
                  width = 3,
                  shinyFiles::shinySaveButton(
                    "save_transcript",
                    "Save plot",
                    "Save file as ...",
                    filetype = list(png = "png", svg = "svg", pdf = "pdf"),
                    style='margin-top:20px'
                  ),
                ),
              ),
              shinycssloaders::withSpinner(shiny::plotOutput("transcript_plot"))
            )
          )
        ),
        shinydashboard::tabItem(
          tabName = "vignette",
          shiny::uiOutput("vignette_ui")
        ),
        shinydashboard::tabItem(
          tabName = "about",
          shiny::fluidRow(
            shinydashboard::box(
              title = "About inspectorORF",
              width = 12,
              status = "primary",
              collapsible = FALSE,
              shiny::div(
                style = "font-size:16px; line-height:1.6;",
                shiny::p(shiny::strong("inspectorORF"), " v", as.character(utils::packageVersion("inspectorORF"))),
                shiny::p("Authors: Eilidh Ward & Isabel Birds"),

                shiny::p("Welcome to the inspectorORF dashboard! This shiny app is an interactive tool for the evaluating and visualising transcripts and ORFs from RNA-Seq, Ribo-Seq and additional *omics data utilising functions from the inspectorORF R package."),
                shiny::p("It is built with Shiny and shinydashboard, and includes tabs for:"),
                shiny::tags$ul(
                  shiny::tags$li("Viewing the vignette for use in R"),
                  shiny::tags$li("Easy to use interface for importing and plotting transcripts and ORFs of interest"),
                  shiny::tags$li("File selection and export tools")
                ),
                shiny::tags$h4(shiny::strong("Local installation")),
                shiny::p("You can install inspectorORF directly from GitHub:"),
                shiny::tags$pre(
                  shiny::tags$code(
                  style = "
                    display: block;
                    font-family: monospace;
                    white-space: pre-wrap;
                    text-indent: 0;
                  ",
                    shiny::HTML(paste("# install.packages('devtools')",
                      "devtools::install_github('aylz83/inspectorORF')",
                      sep = "<br>"
                    ))
                  )
                ),

                shiny::p("See the vignette tab for usage as either an R package or shiny application."),
                shiny::tags$p("Please also see the GitHub respository at ",
                  shiny::tags$a(
                    href = "https://github.com/aylz83/inspectorORF", 
                    "https://github.com/aylz83/inspectorORF",
                    target = "_blank"
                  )
                ),
                shiny::br(),
              ),
              shiny::tags$div(
                style = "font-size:16px; line-height:1.6;",
                shiny::tags$h4(shiny::strong("Example Output")),
                shiny::tags$img(src = "https://raw.githubusercontent.com/aylz83/inspectorORF/main/images/example_plot.png", width = "80%", style = "border:1px solid #ccc; border-radius:8px;")
              ),
              shiny::tags$div(
                style = "font-size:16px; line-height:1.6;",
                shiny::tags$hr(),
                shiny::p(shiny::strong("Acknowledgements:")),
                shiny::p("This package uses several R packages, including shiny, shinydashboard, shinyjs, shinyFiles, reactable, and others.")
              )
            )
          )
        )
      )
    ),
    skin = "purple"
  )
}
